garden_ai/model.py

```diff
@@ -0,0 +1,146 @@
+    ----------
+    creators: List[str]
+        The main researchers involved in producing the Garden. At least one
+        creator is required in order to register a DOI. Personal name format
+        should be: "Family, Given".
+
+    title: str
+        An official name or title for the Garden. This attribute must be set
+        in order to register a DOI.
+
+
+    Recommended Attributes
+    ----------
+
+    description: str
+        A brief summary of the Garden and/or its purpose, to aid discovery by
+        other Gardeners.
+
+
+    Optional Attributes
+    ----------
+    doi: str
+        A garden's doi can be manually set by the user, or generated automatically via the
+        DataCite rest api using the required fields.
+
+    language: str
+        Recommended values are IETF BCP 47, ISO 639-1 language codes, such as:
+        "en", "de", "fr". Defaults to "en".
+
+    subjects: List[str]
+        Subjects, keywords, classification codes, or key phrases pertaining to the Garden.
+
+    Examples
+    --------
+    Creating a new Garden with a ``GardenClient`` instance::
+
+        gc = GardenClient()
+        pea_garden = gc.create_garden(
+            creators=["Mendel, Gregor"], title="Experiments on Plant Hybridization"
+        )
+        pea_garden.creators += ["St Thomas' Abbey"]
+        pea_garden.year = 1863
+        pea_garden.description = '''This Garden houses sophisticated ML pipelines
+                                  for Big Pea Data extraction and classification.
+                                  It consists of a 2-hectare plot behind the monastery,
+                                  and a 30,000-plant dataset.'''
+
+        gc.register(pea_garden) # we can also have this invoke a pydantic "root validator" we define elsewhere
+
+    Notes
+    --------
+    Mendel's work was ignored by the scientific community during his lifetime,
+    presumably due to the lack of a working DOI.
+    To remedy this, if the doi field is unset when registering the
+    garden, we could just build one for the user with the datacite api
+    """
+
+    class Config:
+        """
+        Configure pydantic per-model settings.
+
+        We disable validate_all so that pydantic ignores our custom TempField defaults
+        We enable validate_assignment to make validation occur naturally even after object construction
+        """
+
+        validate_all = False
+        validate_assignment = True  # validators invoked (only?) on assignment
+
+    creators: List[str] = Field(TempField.REQUIRED, min_items=1, unique_items=True)
+    title: str = Field(TempField.REQUIRED)
+    resourceTypeGeneral: str = "Other"  # (or: model, software, service, interactive?)
+    publisher: str = "Garden"
+    year: str = Field(default_factory=lambda: str(datetime.now().year))
+
+    # dois currently only validated against this regex
+    # doi: str = Field(TempField.REQUIRED, alias="identifier", regex=r"^10.\d{4,9}/[-._;()/:A-Z0-9]+$")
```

Comment:
is the commented out code in this Config meant to be here?

Reply:
the regex convenience option that pydantic provided there was causing later validation for the doi to skip for some reason, I'm not sure if it was a pydantic bug or a me bug and I thought I might revisit it later, but also "I'll revisit this later" could be famous last words so I can just take it out